meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClient.java

```diff
@@ -332,15 +333,18 @@ public boolean subscribeStateChanges(ConnectStateChangeListener listener) {
     return true;
   }
 
-  // TODO: add impl and remove UnimplementedException
   @Override
   public boolean subscribeChildChanges(String key, ChildChangeListener listener, boolean skipWatchingNonExistNode) {
+    if (skipWatchingNonExistNode && exists(key) == null) {
+      return false;
+    }
     try {
-      _zkClient.subscribePersistRecursiveWatcher(key, new ChildListenerAdapter(listener));
-    } catch (KeeperException.UnimplementedException e) {
-      LOG.error(e.getLocalizedMessage());
+      _zkClient.subscribePersistRecursiveListener(key, new ChildListenerAdapter(listener));
+    } catch (ZkException ex) {
+      LOG.error("Failed to subscribe ChildChanges for path: " + key, ex);
```

Comment:
Open question, what’s the idea behind catching the exception vs failing out? It seems for subscribe method, we catch zk exception, but for unsubscribe, we don’t.

Reply:
Good point! 
I updated the function. It will return false when it is expected to fail (when path does not exists and 'skipWatchingNonExistNode' is true). It throws exception when failure is not expected.